torchvision/models/_api.py

```diff
@@ -203,19 +204,45 @@ def wrapper(fn: Callable[..., M]) -> Callable[..., M]:
     return wrapper
 
 
-def list_models(module: Optional[ModuleType] = None) -> List[str]:
+def list_models(
+    module: Optional[ModuleType] = None,
+    include_filters: Union[Iterable[str], str, None] = None,
+    exclude_filters: Union[Iterable[str], str, None] = None,
+) -> List[str]:
     """
     Returns a list with the names of registered models.
 
     Args:
         module (ModuleType, optional): The module from which we want to extract the available models.
+        include_filters (str or Iterable[str], optional): Filter(s) for including the models from the set of all models.
+            Filters are passed to fnmatch to match Unix shell-style wildcards. In case of many filters, the results is
+            the union of individual filters.
+        exclude_filters (str or Iterable[str], optional): Filter(s) applied after include_filters to remove models.
+            Filter are passed to fnmatch to match Unix shell-style wildcards. In case of many filters, the results is
+            removal of all the models that match any individual filter.
 
     Returns:
         models (list): A list with the names of available models.
     """
-    models = [
+    all_models = [
         k for k, v in BUILTIN_MODELS.items() if module is None or v.__module__.rsplit(".", 1)[0] == module.__name__
     ]
+    if include_filters is not None:
```

Comment:
I'd suggest

```suggestion
    if include_filters:
```

So that cases like `list_models(include_filters=[])` are equivalent to `list_models()`. (same below).

Most people won't be writing `list_models(include_filters=[])` directly but it's possible to write `list_models(include_filters=my_filter)` where `my_filter` ends up being an empty list.

Reply:
With this one I am not entirely sure, it is an edge case. Default is None, and it means include all. But if not None is passed, it is an explicit Iterable of what should be included. 

The more strict the pattern, the less models you should return, with extreme cases of exact match, which result in a single model, and then one step further is to have an empty list, which for the sake of consistency could return nothing. 

I imagine use case where you produce a list of include filters using set operations, and some of them may produce empty lists. Then I think it may be safer to fail fast (by ending up with no models), instead of suddenly flipping to accepting all possible models.

Please advise :)

Reply:
Thanks @MateuszGuzek , that makes sense. I guess there are cases to be made for both options.

@pmeier @vfdev-5 any thought on that? Basically we're trying to decide whether `list_models(include=[])` should be interpreted as A) "there's no filter, so don't filter anything, return the whole list" or B) as "there's no filter, so there's nothing to include, so we return nothing".

Note that timm uses A (which IMHO is a good reason to go for that):

https://github.com/huggingface/pytorch-image-models/blob/c241081251323dfc5e8dc799d49740c48cc9096f/timm/models/_registry.py#L205

Reply:
Since the feature request in https://github.com/pytorch/vision/issues/6365#issuecomment-1220097910 specifically mentions

> timm has this functionality and it really boosts productivity.

Thus, we should align with what they have, i.e. go for A

Reply:
> Thus, we should align with what they have, i.e. go for A

This is a fair point, I will go for this implementation.